revocation/revocation.go

```diff
@@ -16,49 +16,97 @@
 package revocation
 
 import (
+	"context"
 	"crypto/x509"
 	"errors"
+	"fmt"
 	"net/http"
 	"time"
 
 	"github.com/notaryproject/notation-core-go/revocation/ocsp"
 	"github.com/notaryproject/notation-core-go/revocation/result"
 )
 
-// Revocation is an interface that specifies methods used for revocation checking
+// Revocation is an interface that specifies methods used for revocation checking.
+//
+// Deprecated: Revocation exists for backwards compatibility and should not be used.
+// To perform revocation check, use [ContextRevocation].
 type Revocation interface {
 	// Validate checks the revocation status for a certificate chain using OCSP
 	// and returns an array of CertRevocationResults that contain the results
 	// and any errors that are encountered during the process
 	Validate(certChain []*x509.Certificate, signingTime time.Time) ([]*result.CertRevocationResult, error)
 }
 
+// ValidateContextOptions provides configuration options for revocation checks
+type ValidateContextOptions struct {
+	// CertChain denotes the certificate chain whose revocation status is
+	// been validated. REQUIRED.
+	CertChain []*x509.Certificate
+
+	// AuthenticSigningTime denotes the authentic signing time of the signature.
+	// It is solely used under signing scheme `notary.x509.signingAuthority`.
+	// OPTIONAL.
+	AuthenticSigningTime time.Time
+}
+
+// ContextRevocation is an interface that provides revocation checking with
+// context
+type ContextRevocation interface {
+	// ValidateContext checks the revocation status given caller provided options
+	// and returns an array of CertRevocationResults that contain the results
+	// and any errors that are encountered during the process
+	ValidateContext(ctx context.Context, validateContextOpts ValidateContextOptions) ([]*result.CertRevocationResult, error)
+}
+
 // revocation is an internal struct used for revocation checking
 type revocation struct {
 	httpClient       *http.Client
-	certChainPurpose ocsp.Purpose
+	certChainPurpose x509.ExtKeyUsage
```

Comment:
Why do we need purpose, can we determine it from leaf certificate in certificate chain?

Reply:
This was introduced in the previous PR (https://github.com/notaryproject/notation-core-go/pull/207) suggested by @shizhMSFT and released in beta.1. So it's not a new field in this PR. We just found that the built-in library is already providing `x509.ExtKeyUsage`, so it's better to use it directly.
Providing a `certChainPurpose` and asking the user to set it `explicitly` adds a guard to prevent the caller accidentally passed in a `code signing cert chain` whereas `timestamping` revocation check is expected, and vice versa.

Reply:
Instead of adding `certChainPurpose` can user pass a cert chain validation function? This way they know what they want.

Reply:
I think the `cert chain validation function` should be controlled by notation-core-go? Because we always want to validate the cert chain during revocation check. If the caller passed in a dummy function, they are allowed to skip the cert chain validation.

For this version, since we already have `certChainPurpose` in beta.1 release, I think we could keep it this way? And we will discuss this further in the v2.0 design of course.